src/Interpreters/InterpreterSelectWithUnionQuery.cpp

```diff
@@ -128,24 +128,24 @@ InterpreterSelectWithUnionQuery::InterpreterSelectWithUnionQuery(
     const ASTPtr & query_ptr_, const Context & context_, const SelectQueryOptions & options_, const Names & required_result_column_names)
     : IInterpreterUnionOrSelectQuery(query_ptr_, context_, options_)
 {
-    auto & ast = query_ptr->as<ASTSelectWithUnionQuery &>();
+    ASTSelectWithUnionQuery * ast = query_ptr->as<ASTSelectWithUnionQuery>();
 
     /// Normalize AST Tree
-    if (!ast.is_normalized)
+    if (!ast->is_normalized)
     {
         CustomizeASTSelectWithUnionQueryNormalizeVisitor::Data union_default_mode{context->getSettingsRef().union_default_mode};
         CustomizeASTSelectWithUnionQueryNormalizeVisitor(union_default_mode).visit(query_ptr);
 
         /// After normalization, if it only has one ASTSelectWithUnionQuery child,
         /// we can lift it up, this can reduce one unnecessary recursion later.
-        if (ast.list_of_selects->children.size() == 1 && ast.list_of_selects->children.at(0)->as<ASTSelectWithUnionQuery>())
+        if (ast->list_of_selects->children.size() == 1 && ast->list_of_selects->children.at(0)->as<ASTSelectWithUnionQuery>())
         {
-            query_ptr = std::move(ast.list_of_selects->children.at(0));
-            ast = query_ptr->as<ASTSelectWithUnionQuery &>();
```

Comment:
N.B. This assigns to an object of ASTSelectWithUnionQuery class.

Reply:
At first I was puzzled by how it manages to cast away constness from query_ptr in constructor of IInterpreterUnionOrSelectQuery, but then I realized that this is a const reference to a smart pointer to a non-const AST.... not sure this construct (that is widely used in our code) makes any sense at all.

By the way, how did you notice it? Maybe there is some test case that breaks?

Reply:
No tests were broken, nothing of that sort. I was just analyzing code from another related PR and could not get my head around this assignment. I do not believe we currently have a real problem because of this as the object's memory most probably stays allocated at the time of the assignment (and so does the reference)

Reply:
Hi, I didn't understand the problem here.